### ISSUE TITLE
R6 counter json list to big to manage

### ISSUE THREAD

**Issue by Mr-Martinez** [BurkusCat/r6-operator-counters#155]
**Issue**
Vertical Growth Issue with R6 counters (1500+ lines of code)

**Reason**
With the popularity of R6 Siege, it seems that this game will have many seasons and new characters added. If all R6 character's nodes are in 1 file, it will be harder to manage and track which characters connect.

**Solution**
I would like to propose a rework on how the nodes are created/managed, making future and current characters easier to work with.
By refactoring how the counters are managed/written, it will be easier to add/update/remove nodes

**Comment by BurkusCat** [comment]
I like the idea behind this. Something that is difficult about working with this file is that it is hard to follow which IDs match to which operator.

Do you have any thoughts on how to improve this file? Describing relationships using operator names and breaking down the file would be a massive improvement I think.

**Comment by Mr-Martinez** [comment]
Just the top of my head, I would start with how layout of files/folders. 

First level folders would hold be labeled with seasons (s1, s2,etc) , then second layer would have folders named after the teams the game characters belong to (eg Spetznaz, GIGN, etc), then the third layer would be .js files that will be named after an individual character (eg glaz.js, rook.js).

**Comment by Mr-Martinez** [comment]
I created a quick, non working repo to show off my thoughts: https://github.com/Mr-Martinez/r6-counters-refactored .

Basically just refactoring json code to its own single file/operator.
Then a module that imports each individual .js file/object to create one big json object that the html file can use to iterate its graph.

We can essentially do the same for images as well.

Basically each file will be responsible for handling data for a single operator.

A global enum for making sure that you don't have to memorize node numbers, opting to use names instead. Easier to see what's going on inside files

**Comment by BurkusCat** [comment]
@marcopixel sent me this library that they've been working on: https://github.com/marcopixel/r6operators that contains all the operator icons + operator bio data. Would it be worth using this library in some way to retrieve the bio data for each operator (e.g. attacker/defender, real name etc.) and then in `r6-operator-counters` add the counter data?

I have a question about the method "getHardCounter();", would this return who that operator hard counters, who that operator is hard countered by, or both?

**Comment by Mr-Martinez** [comment]
@marcopixel, this is what I had in mind indeed!

> Would it be worth it using this library...

Yes! Although it really depends... Will @marcopixel update the file with every season of the year and/or allow anyone to help with the project? (I'd like to help out when I'm able to)

Reading https://github.com/marcopixel/r6operators , the heart of the code is 'operators.json' file, and nothing there looks like they rely on each other. Meaning it would be easy for anyone to help out on the project as well if there are times where @marcopixel is not able.

> I have a question about the method "getHardCounter();", would this return who that operator hard counters, who that operator is hard countered by, or both?

In this case the operators are returning who they are hard counter by. 
Although a change in naming convention is needed avoid confusion.
Any ideas?

So far I'm thinking...
"get" vs"retrieve" vs other synonyms.
"HardCounteredby()" vs "HardCounteredFrom()"

Although, with each file having the responsibility of taking care of only one operator, I can as well add code that shows who the operator hard counter as well if you want to add that as well. 

I'm coming from a typescript background, but in my pseudo-code I'm trying to return back one big Class Object with functions that can handle all sorts of goodies for a single operator. In this cases the three options you can pick on the website, 

I wanted feedback on this style in addition to the folder structure. Is there a benefit of 3 layers deep, or two layers deep when using gulp? I have to read the Gulp API documentation properly, but to my understanding gulp can grab data from a single file, then add it to a completely new file essentially combining all data? So in my mind I'm thinking gulp can combine outputs of all files in lets say the GIGN folder making a GIGN json file. 

I basically want to make it easier for future use, lets say someone wants to make a filter list for organizations operators belong to.

**Comment by marcopixel** [comment]
I'm intending to support it as long i've got interrest into the game and i'm also open to add contributors to the repo :)

**Comment by BurkusCat** [comment]
@Mr-Martinez Potentially GetCountersTo() and GetCountersFrom()? Unsure. Maybe these methods could take in an array of enums of what levels of counter to return?

Folder structure wise, would it make sense to divide operators up by Year then Season? As an operator may come out in Year 5 that is apart of an organisation already previously used. I think as long as we get the main benefit of the data being split up. Even splitting up each operator into their own file before combining would be of large benefit.

**Comment by Mr-Martinez** [comment]
Oh that's nice!
I'll get on it now.

**Comment by marcopixel** [comment]
That's also how i've intended it on r6operators when i'll decide to include additional assets like front shots & gadgets. One main folder called operators and subfolders for each operator, each of them including the JSON metadata and the assets.

Seems like the most reasonable & clean approach for this problem in my opinion.
